[LLDB] Fix 64 bit support for CIE and FDE handling in DWARFCallFrameInfo

agontarek opened the pull request:
- Introduced a new helper function `IsCIEMarker` to determine if a given `cie_id` indicates a CIE (Common Information Entry) or FDE (Frame Description Entry).
- New helper function can now correctly identify both 32-bit and 64-bit CIE based on the DWARF specifications.
- Updated the `ParseCIE` and `GetFDEIndex` methods to utilize the new helper function for improved clarity and correctness in identifying CIE and FDE entries.
- Replaced direct comparisons with `UINT32_MAX` and `UINT64_MAX` with `std::numeric_limits` for better readability.

agontarek committed:
[LLDB] Refactor CIE and FDE handling in DWARFCallFrameInfo

- Introduced a new helper function `IsCIEMarker` to determine if a given `cie_id` indicates a CIE (Common Information Entry) or FDE (Frame Description Entry).
- New helper function can now correctly identify both 32-bit and 64-bit CIE based on the DWARF specifications.
- Updated the `ParseCIE` and `GetFDEIndex` methods to utilize the new helper function for improved clarity and correctness in identifying CIE and FDE entries.
- Replaced direct comparisons with `UINT32_MAX` and `UINT64_MAX` with `std::numeric_limits` for better readability.

clayborg reviewed (approved):
Looks good!

clayborg commented on lldb/source/Symbol/DWARFCallFrameInfo.cpp:
```diff
@@ -147,6 +149,27 @@ GetGNUEHPointer(const DataExtractor &DE, lldb::offset_t *offset_ptr,
+
+  // DWARF type
+  if (is_64bit)
+    return cie_id == std::numeric_limits<uint64_t>::max();
+
+  return cie_id == std::numeric_limits<uint32_t>::max();
```
Do we want to make a `constexpr` with variable names for this instead of using magic numbers? I was thinking something like:
```
static constexpr uint32_t kCIEMarker32 = std::numeric_limits<uint32_t>::max();
static constexpr uint64_t kCIEMarker64 = std::numeric_limits<uint64_t>::max();
```
And then use these everywhere that we were using magic numbers?

dmpots replied:
We already have these constants in the llvm Dwarf.h header

```
const uint32_t DW_CIE_ID = UINT32_MAX;
const uint64_t DW64_CIE_ID = UINT64_MAX;
```

We can use them from there
https://github.com/dmpots/llvm-project/blob/main/llvm/include/llvm/BinaryFormat/Dwarf.h#L97

clayborg commented on lldb/source/Symbol/DWARFCallFrameInfo.cpp:
```diff
@@ -147,6 +149,27 @@ GetGNUEHPointer(const DataExtractor &DE, lldb::offset_t *offset_ptr,
+
+  // DWARF type
+  if (is_64bit)
+    return cie_id == std::numeric_limits<uint64_t>::max();
```
```
return cie_id == kCIEMarker32;
```

clayborg commented on lldb/source/Symbol/DWARFCallFrameInfo.cpp:
```diff
@@ -147,6 +149,27 @@ GetGNUEHPointer(const DataExtractor &DE, lldb::offset_t *offset_ptr,
+  if (is_64bit)
+    return cie_id == std::numeric_limits<uint64_t>::max();
+
+  return cie_id == std::numeric_limits<uint32_t>::max();
```
return cie_id == kCIEMarker64;

clayborg commented on lldb/source/Symbol/DWARFCallFrameInfo.cpp:
```diff
@@ -283,7 +306,7 @@ DWARFCallFrameInfo::ParseCIE(const dw_offset_t cie_offset) {
   uint32_t length = m_cfi_data.GetU32(&offset);
   dw_offset_t cie_id, end_offset;
-  bool is_64bit = (length == UINT32_MAX);
+  bool is_64bit = (length == std::numeric_limits<uint32_t>::max());
```
```
bool is_64bit = (length == kDWARF64LengthPrefix);
```

dmpots replied:
Can use
```
DW_LENGTH_DWARF64
````

from here
https://github.com/dmpots/llvm-project/blob/5c17af419ec7e9c823a1ae9287baeed0d27336ce/llvm/include/llvm/BinaryFormat/Dwarf.h#L56C3-L56C20

agontarek replied:
Done.

clayborg commented on lldb/source/Symbol/DWARFCallFrameInfo.cpp:
```diff
@@ -470,7 +494,7 @@ void DWARFCallFrameInfo::GetFDEIndex() {
     dw_offset_t cie_id, next_entry, cie_offset;
     uint32_t len = m_cfi_data.GetU32(&offset);
-    bool is_64bit = (len == UINT32_MAX);
+    bool is_64bit = (len == std::numeric_limits<uint32_t>::max());
```
```
bool is_64bit = (len == kDWARF64LengthPrefix);
```

dmpots commented:
> I do not know the proper way to test this change and am open to recommendations on how to add a test

We should be able to add a test to DWARFCallFrameInfoTest using an ELF YAML file that has a 32-bit/64-bit CIE.

Here is an example where I added a minimal CIE entry for testing.

https://github.com/clayborg/llvm-project/pull/43/files#diff-203f02495ba27888f5299515d238956d6abc1efe8fdc07d89a6347a84ff8d016

dmpots reviewed (changes requested):
I think we should add a test for this. Can help you with the YAML if needed.

agontarek committed:
[LLDB] Update DWARF CIE marker checks to use LLVM constants

This commit refactors the CIE marker checks in DWARFCallFrameInfo.cpp to utilize LLVM's predefined constants for DWARF CIE IDs.

agontarek committed:
[LLDB] Add tests for handling invalid FDE entries and valid CIE markers

This commit introduces new unit tests to verify the handling of invalid Frame Description Entries (FDEs) with invalid CIE ID values for both DWARF32 and DWARF64 formats. Additionally, it includes tests for valid CIE markers in both eh_frame and debug_frame formats, ensuring that the unwind plans are correctly generated or returned as null when encountering invalid entries.

walter-erquinigo reviewed (approved):
lgtm

dmpots reviewed (approved):
LGTM. Thanks for the changes and adding the test!

> I do not know the proper way to test this change and am open to recommendations on how to add a test.

We can update the commit message now that we have tests!

Question for my understanding. What would be the behavior of the tests without your fix? Would it crash in that case?

agontarek committed:
Merge branch 'main' into agontarek/DWARFCallFrameInfo
